### ISSUE TITLE
ImmutableArray<T> struct implements IEquatable<T> interface but does not stick to the rules of properly implementing value semantics

### ISSUE THREAD

**Issue by oliverzick** [dotnet/runtime#77183]
### Description

The `ImmutableArray<T>` struct implements the `IEquatable<T>` interface but does not stick to the rules of implementing value semantics acoording to the [IEquatable<T>](https://learn.microsoft.com/en-us/dotnet/api/system.iequatable-1?view=net-6.0) interface.

When comparing two instance of a `ImmutableArray<T>` both instance are considered equal only when both instances refer to the same array instance. Instead both instances should be considered equal when the contents of the contained collections are equal.

### Reproduction Steps

```
using System.Collections.Immutable;

var collection = Enumerable.Range(1, 5);

var instance1 = collection.ToImmutableArray();
var instance2 = collection.ToImmutableArray();

// Result = False ---> Wrong result for equality in terms of value semantics
var result = instance1.Equals(instance2); 

Console.WriteLine($"Result: {result}"); // Outputs "Result: False"
```

### Expected behavior

```
using System.Collections.Immutable;

var collection = Enumerable.Range(1, 5);

var instance1 = collection.ToImmutableArray();
var instance2 = collection.ToImmutableArray();

// Result = True ---> Should be true because both instances represent a collection of integers from 1 to 5 which are equal in terms of value semantics defined by IEquatable<T> interface
var result = instance1.Equals(instance2);

Console.WriteLine($"Result: {result}"); // Outputs "Result: True"
```

### Actual behavior

Already described in Description and Reproduction Steps sections.

### Other information

The `IEquatable<T>` interface should by properly implemented according to its specification to ensure value semantics of `ImmutableArray<T>`. This also includes desired behavior of Equals and GetHashCode methods according to interface specification.

**Comment by Clockwork-Muse** [comment]
> The `ImmutableArray<T>` struct implements the `IEquatable<T>` interface but does not stick to the rules of implementing value semantics acoording to the [IEquatable](https://learn.microsoft.com/en-us/dotnet/api/system.iequatable-1?view=net-6.0) interface.

`IEquatable<T>` doesn't require types implement deep value semantics.  It implies no contract other than some loosely defined "equality" that's up for interpretation by types that implement it.  It's not completely uncommon for either shallow equality or ignored properties to be common in implementations in C# or other languages.

**Comment by eiriktsarpalis** [comment]
And even if we wanted to implement structural equality at this point, it would most likely be a breaking change.

**Comment by RikkiGibson** [comment]
I think the typical guidance here is to use SequenceEqual, but the behavior can be a little troubling if you want to deeply compare records which contain collections, for example. (Records use EqualityComparer<TField>.Default to compare their fields.) [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwBYBuYsyvAOgBlYBHeo4gNwmTlkecgF5yMAKYB3cgCU8ACimz2AZwQAeOTgB85AN4SZ8nMpPqtsBPqMiAvgEpyT7v0HIcY47IXnVUBraeoY+pv7klprWtuQOzq7EDHgAnIrCzACiHKgQADZq6TiOjrTkAPTl5ABi+WqSyWmZMKhgksgQwHmSzADKkjmSMADGktm5eeksAJIIkmBqcEI4zLPzaiVllTV1DTwkqVNZOXVFmxVVtQV7VE0tbR1dPf2DI2MnkxlrCwDapAC6zAAcpINJIACbfRbLVZzX4A4GguaQuEbUoXcgAFTQe2IyEkwwA9shwfIlFEdPooaU8QTiaSdIoojFyCCwSj1qUgA===)

```cs
using System;
using System.Collections.Generic;
using System.Linq;

var r1 = new R1(new List<R2> { new R2(new List<int> { 1 }) });
var r2 = new R1(new List<R2> { new R2(new List<int> { 1 }) });

Console.WriteLine(r1.Equals(r2)); // False
Console.WriteLine(Enumerable.SequenceEqual(r1.Items, r2.Items)); // False

Console.WriteLine(r1.Equals(r2)); // False
Console.WriteLine(Enumerable.SequenceEqual(r1.Items[0].NestedItems, r2.Items[0].NestedItems)); // True

record R1(List<R2> Items);
record R2(List<int> NestedItems);
```

That said, I'm not sure the problem is really "fixable", except that if you want these kinds of equality semantics, to take care to use SequenceEqual or equivalent methods when comparing collections, and perhaps do the same when you implement IEquatable.Equals.

**Comment by Sergio0694** [comment]
(Partially unrelated) Just wanted to say that while I agree that of course changing the behavior of `IEquatable<T>` here is definitely a breaking change and not doable, the current behavior is indeed extremely unfortunate when using lots of records like @RikkiGibson mentioned. This is extremely common when writing source generators (as incrementa generators require you to define models for your incremental steps), and the current situation is that the second you have even a single record parameter that's an immutable array, you just fall off a cliff and have to go all the way down to implementing equality manually for the whole record type, which is a pain. It would be nice if there was _some_ way to improve this scenario, as it's arguably pretty common and unfortunately (1) extremely verbose and (2) extremely error prone, as if you don't know this behavior or just forget when refactoring, you'll silently break your entire incremental pipeline after that step 😅

**Comment by eiriktsarpalis** [comment]
@Sergio0694 Perhaps exposing an EqualityComparer combinator could help in that regard?
```C#
public static class EqualityComparer
{
     public static IEqualityComparer<TList> CreateListComparer<TList, TElement>(IEqualityComparer<TElement> elementComparer) 
        where TList : IList<TElement> { }
}
```

Related to https://github.com/dotnet/runtime/issues/33873 and https://github.com/dotnet/runtime/issues/19644

**Comment by oliverzick** [comment]
As @RikkiGibson and @Sergio0694 already mentioned the current equatable behavior of the `ImmutableArray<T>` struct - whether this was intended or not - is a pitfall, unfortunately.

At least I have experienced this on heavily utilizing records (e.g. when working with DDD or object-functional techniques that involve a lot of immutability and types that come with value semantics). So I was introducing new records representing composites that own their contained collections. Then I was looking for some out-of-the-box immutable collection and the `ImmutableArray<T>` was the only one I could find which is implementing the `IEquatable<T>` interface.
After replacing a custom immutable collection type that comes with value semantics the corresponding specs and tests were broken. So I digged a bit depper and detected that the `ImmutableArray<T>` was actually not what I thought it is according to the name of the type and implementation of the `IEquatable<T>` interface. It looks like @Sergio0694 came across a similar situation. :sweat_smile:

@eiriktsarpalis I know that changing this behavior afterwards would be a breaking change. In my opinion it would be great if there was an out-of-the-box immutable collection or bare enumerable implementation that comes with value semantics and can be used within records.

**Comment by Sergio0694** [comment]
> "it would be great if there was an out-of-the-box immutable collection or bare enumerable implementation that comes with value semantics and can be used within records."

You know what, that's... Actually a very good idea. I reckon using something like that might very well make authoring records much much simpler, as there would be no more a need to handroll custom comparers every single time 🤔

I've hacked together really quick just to play around:

```
public readonly struct EquatableArray<T> : IEquatable<EquatableArray<T>>
    where T : IEquatable<T>
{
    private readonly T[] array;

    public bool Equals(EquatableArray<T> array)
    {
        return AsImmutableArray().SequenceEqual(array.AsImmutableArray());
    }

    public override bool Equals([NotNullWhen(true)] object? obj)
    {
        return obj is EquatableArray<T> array && Equals(this, array);
    }

    public override int GetHashCode()
    {
        if (this.array is not T[] array)
        {
            return 0;
        }

        HashCode hashCode = default;

        foreach (T item in array)
        {
            hashCode.Add(item);
        }

        return hashCode.ToHashCode();
    }

    public ImmutableArray<T> AsImmutableArray()
    {
        return Unsafe.As<T[], ImmutableArray<T>>(ref Unsafe.AsRef(in this.array));
    }

    public static EquatableArray<T> FromImmutableArray(ImmutableArray<T> array)
    {
        EquatableArray<T> array2 = default;

        Unsafe.AsRef(in array2.array) = Unsafe.As<ImmutableArray<T>, T[]>(ref array);

        return array2;
    }

    public static implicit operator EquatableArray<T>(ImmutableArray<T> array)
    {
        return FromImmutableArray(array);
    }

    public static implicit operator ImmutableArray<T>(EquatableArray<T> array)
    {
        return array.AsImmutableArray();
    }

    public static bool operator ==(EquatableArray<T> left, EquatableArray<T> right)
    {
        return left.Equals(right);
    }

    public static bool operator !=(EquatableArray<T> left, EquatableArray<T> right)
    {
        return !left.Equals(right);
    }
}
```

And you 'd use it like so:

```csharp
record MyRecord(ImmutableArray2<string> Names, ImmutableArray2<int> Numbers);

ImmutableArray<string> names = ...;
ImmutableArray<int> numbers = ...;

MyRecord myRecord = new(names, numbers); // Implicit conversion

names = myRecord.Names;
numbers = myRecord.Numbers; // Also implicit conversion
```

I might actually try this out in a branch of the MVVM Toolkit and see how this goes 😄
It _would_ seem like this would simplify a _ton_ of code!

**Comment by GSPP** [comment]
@oliverzick @Sergio0694 I had that problem as well. I wrote myself a struct wrapper that confers value semantics to any list. Something like:

```csharp
struct ValueList<T> : IList<T>
{
    IList<T> inner;
    IEqualityComparer<T> elementComparer;

    //...
}

static class ValueList
{
    public static ValueList<T> Create<T>(IList<T> inner) => ...
}

var bytes = ValueList.Create(myByteArray);
```

That type can ease a lot of pain, especially in the context of LINQ. The main drawback is that the `ValueList` type leaks into the public surface area of APIs.

**Comment by Sergio0694** [comment]
Yup this is not something I'd recommend for public APIs. In my scenario though I was just referring to models in incremental generator steps, which are all internal anyway, so that's not an issue. @eiriktsarpalis that experiment was really a success both in the MVVM Toolkit and ComputeSharp, and I've also discovered yesterday that @jkoritzinsky has also been using an approach much like this in all the various COM interop generators as well. I wonder if this same idea could also help you simplify things in the System.Text.Json generators? Just really feels like a much simpler approach for incremental models 😄

**Comment by mrpmorris** [comment]
I came here to report the same issue. 

Without deep comparison, the automatic `Equals` implemented by `record` classes doesn't tell us two records are equal, it only tells us if they are equal *and* they share a common ancestor that created the `Immutable*` collection, and that's not really very useful.

The same problem exists for all the `Immutable*` types.

**Comment by Sergio0694** [comment]
Just wanted to mention that this issue keeps popping up very often whenever I see people writing source generators (myself included). I had to copy-paste my `EquatableArray<T>` type in over 4 different projects for the same reason, and it's just not ideal. This feels like a common enough issue for source generator authors that it'd be nice if this could either be addressed in `ImmutableArray<T>` itself, or maybe we should propose a separate API that could be used in these cases? I know some other generator authors (eg. @jkoritzinsky) have alternative solutions to this, but it just seels like everyone is reinventing ways to solve the same issue, so maybe just having a proper solution available for everyone would be better? 🤔

**Comment by eatdrinksleepcode** [comment]
Ran into this today, but in my case the lack of structural equality by default didn't just break the caching in the incremental pipeline. My generation function is doing a `GroupBy` on an element of the model that contains an `ImmutableArray<ISymbol>`, which results in the generator producing incorrect output. Fortunately my tests caught it, but this is definitely a pit of failure for generators (one of many, unfortunately).

**Comment by ChrML** [comment]
How about special-casing ImmutableArray in the default equality implementation of records? It's not ideal, but just throwing in the idea.

**Comment by Sergio0694** [comment]
> "My generation function is doing a GroupBy on an element of the model that contains an ImmutableArray<ISymbol>, which results in the generator producing incorrect output."

@eatdrinksleepcode Just read that message — this is a side point to the rest of this conversation, but definitely **do not** capture symbols in your incremental models, ever. It doesn't matter whether you use a custom comparer or whatever to fix incrementality, symbols should just never flow across incremental steps because they cause compilations to be rooted across incremental runs.

**Comment by mrpmorris** [comment]
@Sergio0694 aren't they just data structures?

What is the alternative?

**Comment by Joe4evr** [comment]
@mrpmorris Create your own type(s) that capture the symbol info you want and copy those over.

**Comment by eiriktsarpalis** [comment]
I find that [this article](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md#incremental-generators) captures good guidance on how to write good incremental source generators.

**Comment by mrpmorris** [comment]
@Joe4evr that's what I hoped you'd say :)

**Comment by drdamour** [comment]
> Yup this is not something I'd recommend for public APIs.

why?

**Comment by eiriktsarpalis** [comment]
FWIW I ended up defining a bunch of equatable collection types for my own pet project source generators:

https://github.com/eiriktsarpalis/typeshape-csharp/tree/main/src/TypeShape.Roslyn/IncrementalTypes

**Comment by ThadHouse** [comment]
+1 to fixing this. Requiring 99% of uses of source generators to use a custom type to work at all is a terrible user experience.

**Comment by tannergooding** [comment]
None of the built-in collection types do equality across the entire sequence of elements for `IEquatable<T>` and it has never been part of the contract of `IEquatable<T>` to behave as such. `IEquatable<T>` in no way requires a specific behavior for value types, reference types, etc. The only thing `IEquatable<T>` requires is in relation to `Equals(object obj)` and `GetHashCode()` where it requires some them to behave consistently across the 3 to ensure that hash sets and other collection types work as expected.

It is an explicit design decision for them to be this way, going back to .NET Framework 1.0. It is non-desirable, potentially very expensive, and may even have security concerns/implications to do element-wise equality by default. -- Security concerns can come in from potentially unbounded or unknown processing time which can lead to things like DDoS opportunities or the like. It can lead to catastrophic performance when used with other collections, dictionaries, etc.

Operations like comparing or formatting collections should do the simple thing by default, because these operations can be expected to be frequently called. You should then have separate APIs, such as `SequenceEquals`, when you intentionally want to do element-wise operations between collections.

This is highly unlikely to be changed, but a different API might be possible if there is sufficient justification.

**Comment by RikkiGibson** [comment]
I am interested in the possibility of 'IStructuralEquatable'. I would want for records to be able to easily use that interface for equality checks on fields, and perhaps to add automatic implementation of the interface on record types as well.

However, I also feel like I have seen relatively few requests/upvotes to address the "half-structural" equality that records have in practice. So there's a lot of other features I would want to do first.

**Comment by GauteAtSemine** [comment]
TL;DR: I thought c# had gotten inherent deep value-equality through records + immutable-collections. In fact custom-collections is required? Please help or fix somehow.

While writing equals-assertion in a test, instead of using the assertion-library's slow, 'magic' value-comparison I wanted to utilise fast, well-defined behaviour in c# records. (Better to require c# knowledge.)
To get 'free/no-code' deep comparisons the plan was to use records all the way down. At first mindlessly stumbled with properties that were regular arrays. Then took the immutability and value-equality of records as a que and tried immutable-array. I was surprised to discover it only did reference-equals.
Then reasoned: 
	No collection with value-comparison exist -> Have to create custom collection to override `Equals`. 
	Don't want to wrap immutable-array to keep no-allocations-benefit -> Must extend it -> Oh no, it is sealed! -> came here.
What to do?

**Comment by Grastveit** [comment]
TL;DR: I thought c# had gotten inherent deep value-equality through records + immutable-collections. 
In fact custom-collections is required? With records already done it seems native deep value-equality-support is low-hanging fruit...

While writing equals-assertion in a test, instead of using the assertion-library's slow, 'magic' value-comparison I wanted to utilise fast, well-defined behaviour in c# records. (Better to require c# knowledge.)
To get 'free/no-code' deep comparisons the plan was to use records all the way down. At first mindlessly stumbled with properties that were regular arrays. Then took the immutability and value-equality of records as a queue and tried immutable-array. Was surprised to discover it only did reference-equals.
Then reasoned: 
	No collection with value-comparison exist -> Have to create custom collection to override `Equals`. 
	Don't want to wrap immutable-array to keep no-allocations-benefit -> Must extend it -> Oh no, it is sealed! -> came here. I think many developers could benefit from an easy-to-find solution to this.

**Comment by thomhurst** [comment]
Surprised at this. Has anyone created a package for an equatable collection?

**Comment by eiriktsarpalis** [comment]
> Has anyone created a package for an equatable collection?

I've published [this package](https://www.nuget.org/packages/TypeShape.Roslyn) that includes a number of [immutable equatable collections](https://github.com/eiriktsarpalis/typeshape-csharp/tree/main/src/TypeShape.Roslyn/IncrementalTypes) that I use in my source generators. Variations of the same types are also used by [generators in this repo](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/SourceGenerators/ImmutableEquatableArray.cs).

**Comment by charlesroddie** [comment]
Why not introduce a warning when an `ImmutableArray` is cast to `IEquatable<'T>`?

So far this thread just has that:
- There is a pitfall because `IEquatable<'T>` suggests that equality is structural and will therefore cause bugs if users make this assumption.
- Changing this to structural equality, or removing the interface, would be breaking changes and seem to be ruled out.

A warning would remove the pitfall without any breaking changes. Even if it is off by default, it would allow users to remove the danger of using this malfunctioning interface by mistake.

**Comment by tannergooding** [comment]
`structural equality` is not the same as `content equality` or `sequence equality` or whatever other terminology you want to define. That is, `structural equality` does not mean that it deeply compares all members using `structural equality`.

Rather, `reference equality` is the default for `reference types` and simply means that the reference is compared, not the fields of the type. While `structural equality` is the default for `value types` and simply means that `Equals` is called on each field within the struct.

Thus, while comparing two `T[]` uses reference equality by default and comparing two `struct S { T[] field; }` uses `structural equality` by default, the latter is going to functionally behave the same as the former because the struct only contains a single field that is itself a reference type. This is how .NET has always worked going back to v1.0 over 20 years ago.

This is likewise exactly how types like `Span<T>` work, how some theoretical `ValueList<T>` would be expected to work, etc. We expose a separate set of APIs such as `SequenceEquals` where relevant that are appropriate for doing the type of deeper equality on collection types that's being asked about in this thread. -- It might be viable to request some new `ISequenceEquatable` interface that allows access to such a `SequenceEquals` member so that these types of comparisons can be done efficiently (which is a bit more direct and explicitly than simply checking if it implements `IEnumerable`, which may miss some things and may not be valid for all collection types).

**Comment by charlesroddie** [comment]
There are cases where reference equality is useful but these are extremely small in domain modelling. It's a 0.01% case among the code that I am familiar with (which is biased to be predominantly F#, where modelling of domains is prominent). Your "not the norm" includes records and tuples but also strings. Add structural equality on lists, DUs, and record types and you have the F# world in which no one needs to bother with reference equality. The shallow vs deep issue and the content/structural/sequence distinctions go away assuming everything inside has structural equality.

Having equality (semantics) depend on whether something is a struct or a reference type (implementation details) is a horrible idea and something dotnet1.0 assumptions 

People want an ImmutableArray to model a sequence of elements, and two such sequences are equal if they are sequence-equal. The difference between shallow and deep

**Comment by tannergooding** [comment]
Such equality is not a viable default in many (if not most) domains due to the potential for security issues, such as DDoS style attacks. It’s something that already exists has to be explicitly considered and specially handled for strings and which causes numerous potential for issues to otherwise unassuming users.

you ultimately want to support multiple types of equality and need them to be differentiated by some form of contract or interface.

IEquatable is, by design, intended for allowing use of types with collections and sets. This in part also implies having bounded defaults so that otherwise basic operations do not present pita of failure by default.

Sequence equality is a different concept that may warrant its own separate interface
